mlflow/utils/_capture_modules.py

```diff
@@ -35,18 +35,41 @@ class _CaptureImportedModules:
     `builtins.__import__` and `importlib.import_module`.
     """
 
-    def __init__(self):
+    def __init__(self, record_full_module=False):
         self.imported_modules = set()
         self.original_import = None
         self.original_import_module = None
+        self.record_full_module = record_full_module
 
     def _wrap_import(self, original):
         @functools.wraps(original)
         def wrapper(name, globals=None, locals=None, fromlist=(), level=0):
             is_absolute_import = level == 0
-            if is_absolute_import:
-                self._record_imported_module(name)
-            return original(name, globals, locals, fromlist, level)
+            if not self.record_full_module:
+                if is_absolute_import:
+                    self._record_imported_module(name)
+
+            result = original(name, globals, locals, fromlist, level)
+
+            if self.record_full_module:
+                if is_absolute_import:
+                    parent_modules = name.split(".")
+                else:
+                    parent_modules = globals["__name__"].split(".")
+                    if level > 1:
+                        parent_modules = parent_modules[: -(level - 1)]
+
+                if fromlist:
+                    for from_name in fromlist:
+                        full_modules = parent_modules + [from_name]
+                        full_module_name = ".".join(full_modules)
+                        if full_module_name in sys.modules:
+                            self._record_imported_module(full_module_name)
+                else:
+                    full_module_name = ".".join(parent_modules)
+                    self._record_imported_module(full_module_name)
```

Comment:
How much overhead does this add? An import statement is called many times. A small delay can sum up to a large delay. Not a blocker, but just curious.

Reply:
> An import statement is called many times

But because of python module caching, only the first importing execution triggers the hook.

Reply:
Is it possible to measure the difference between `infer_code_paths=True` and `infer_code_paths=False`?

Reply:
or you can insert `print` to see how often it's called.

Reply:
I think it's worth refactoring. The performance is important.